File: backend/src/models/posts.ts
import { z } from 'zod'
import { Result } from './result'
import { ApiError } from './errors'

export type Target = 'mastodon' | 'bluesky' | 'twitter' | 'linkedin'

export const NewPostRequest = z.object({
  content: z.string().min(1).max(1000),
  targets: z.array(z.enum(['mastodon', 'bluesky', 'twitter', 'linkedin'])).optional(),
  link: z.string().optional(),
  language: z.string().optional(),
  cleanupHtml: z.boolean().optional(),
  images: z.array(z.string()).optional()
})

export type NewPostRequest = z.infer<typeof NewPostRequest>

export type NewPostResponse =
  | NewMastodonPostResponse
  | NewBlueSkyPostResponse
  | NewTwitterPostResponse
  | NewRssPostResponse

export type NewBlueSkyPostResponse = {
  module: 'bluesky'
  uri: string
  cid?: string
}

export type NewMastodonPostResponse = {
  module: 'mastodon'
  uri: string
}

export type NewRssPostResponse = {
  module: 'rss'
  uri: string
}

export type NewTwitterPostResponse = {
  module: 'twitter'
  id: string
}

export type CreatePostFunction = (
  post: NewPostRequest
) => Promise<Result<NewPostResponse, ApiError>>

File: backend-kotlin/src/main/kotlin/com/alexn/socialpublish/models/Posts.kt
package com.alexn.socialpublish.models

import kotlinx.serialization.Serializable

enum class Target {
    MASTODON,
    BLUESKY,
    TWITTER,
    LINKEDIN
}

@Serializable
data class NewPostRequest(
    val content: String,
    val targets: List<String>? = null,
    val link: String? = null,
    val language: String? = null,
    val cleanupHtml: Boolean? = null,
    val images: List<String>? = null
) {
    fun validate(): ValidationError? {
        if (content.isEmpty() || content.length > 1000) {
            return ValidationError(
                status = 400,
                errorMessage = "Content must be between 1 and 1000 characters"
            )
        }
        return null
    }
}

sealed class NewPostResponse {
    abstract val module: String
}

@Serializable
data class NewBlueSkyPostResponse(
    override val module: String = "bluesky",
    val uri: String,
    val cid: String? = null
) : NewPostResponse()

@Serializable
data class NewMastodonPostResponse(
    override val module: String = "mastodon",
    val uri: String
) : NewPostResponse()

@Serializable
data class NewRssPostResponse(
    override val module: String = "rss",
    val uri: String
) : NewPostResponse()

@Serializable
data class NewTwitterPostResponse(
    override val module: String = "twitter",
    val id: String
) : NewPostResponse()
